core/jobs/batch_jobs/exp_validation_jobs.py

```diff
@@ -0,0 +1,311 @@
+            for answer_group in answer_groups:
+                if len(answer_group.training_data) > 0:
+                    state_names.append(state_name)
+                    # The below break is necessary to avoid duplicated
+                    # state names in the list.
+                    break
+        return state_names
+
+    def filter_curated_explorations(
+        self, exp_and_opp_models: tuple[str, dict[str, list[object]]]):
+        """Returns whether the exp model is curated or not.
+
+        Args:
+            exp_and_opp_models: tuple. The pair of exp id and models.
+
+        Returns:
+            bool. Returns whether the exp model is curated or not.
+        """
+        exp_models_list = list(exp_and_opp_models[1]['exp_model'])
+        opp_models_list = list(exp_and_opp_models[1]['exp_opportinity_model'])
+        return len(exp_models_list) == 1 and len(opp_models_list) == 1
+
+    def get_exploration_from_models(
+        self, exp_and_opp_models: tuple[str, dict[str, list[object]]]):
+        """Returns the exploration domain object from the curated
+        exploration model.
+
+        Args:
+            exp_and_opp_models: tuple. The pair of exp and opportunity models.
+
+        Returns:
+            Exploration. The exploration domain object.
+        """
+        exp_model = list(exp_and_opp_models[1]['exp_model'])[0]
+        return exp_fetchers.get_exploration_from_model(exp_model)
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns PCollection of details of explorations which are invalid.
+
+        Returns:
+            PCollection[JobRunResult]. A PCollection of details of explorations
+            which are invalid.
+        """
+        all_explorations: beam.PCollection = (
+            self.pipeline
+            | 'Get all ExplorationModels' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all(include_deleted=False))
+            | 'Create key-value pairs for exp models' >> beam.Map(
+                lambda exp_model: (exp_model.id, exp_model))
+        )
+
+        all_exp_opportunities: beam.PCollection = (
+            all_explorations
+            | 'Get all ExplorationOpportunitySummaryModels' >>
+                ndb_io.GetModels(
+                    opportunity_models.ExplorationOpportunitySummaryModel
+                        .get_all(include_deleted=False))
+            | 'Create key-value pairs for opportunity models' >> beam.Map(
+                lambda exp_opportunity_model: (
+                    exp_opportunity_model.id, exp_opportunity_model))
+        )
+
+        curated_explorations: beam.PCollection = (
+            ({
+                'exp_model': all_explorations,
+                'exp_opportinity_model': all_exp_opportunities
+            })
+            | 'Combine the PCollections' >> beam.CoGroupByKey()
+            | 'Filter curated explorations' >> beam.Filter(
+                self.filter_curated_explorations)
```

Comment:
Maybe you can also reformat the pcoll returned by `CoGroupByKey` so that it is just a dictionary with key and model value. Because now you need to do `list(exp_and_opp_models[1]['exp_model'])` above, yo should modify it here so that you can do just `exp_and_opp_models['exp_model']` above to get the model.

Reply:
I just noticed that we don't need to convert into list when we use `CoGroupByKey`. I have modified this functionality now. PTAL. Thanks!